src/fal/dag.py

```diff
@@ -0,0 +1,129 @@
+from dataclasses import dataclass, field
+from os import path
+from typing import List,TypeVar, Dict
+from faldbt.project import DbtModel 
+from pathlib import Path
+import re
+
+T = TypeVar("T", bound="FalScript")
+
+@dataclass(frozen=True)
+class FalScript:
+    model: DbtModel
+    path: str
+    dependencies: List[T] = field(default_factory=list)
+
+    def __hash__(self):
+        return self.path.__hash__()
+
+    def __repr__(self):
+        return "FalScript(" + self.path.__str__() + ")"
+
+    def parse(self) -> List[str]:
+        """
+        Parses the python file to identify any dependencies
+        """
+        with open(self.path) as file:
+            lines = file.readlines()
+            ## TODO: do this search in parallel
+            dependency_names = []
+            for line in lines:
+                match = re.search('ref\("(.*)"\)', line)
+                if match:
+                    group = match.group(1)
+                    dependency_names.append(group)
+            return dependency_names    
+    
+    def addDependency(self, dependency: T):
+        if (dependency not in self.dependencies):
+            self.dependencies.append(dependency)
+
+    def exec(self, ref, context, source):
+        """
+        Executes the script
+        """
+        with open(path) as file:
+            a_script = file.read()
+            exec(
+                a_script,
+                {
+                    "ref": ref,
+                    "context": context,
+                    "source": source,
+                },
+            )
+
+@dataclass(frozen=True)
+class UniqueKey:
+    model: DbtModel
+    script_path: Path
+
+class ScriptGraphBuilder:
+    modelNameToModelLookup: Dict[str, DbtModel] = {}
+    modelToScriptLookup: Dict[str, List[Path]] = {}
```

Comment:
`modelNameToModelLookup` and `modelToScriptLookup` could be a single lookup map with tuples. Not that it is necessarily better.